community/kernel/src/main/java/org/neo4j/kernel/impl/store/RecordStore.java

```diff
@@ -458,37 +451,45 @@ public abstract void processRelationshipGroup( RecordStore<RelationshipGroupReco
         private <R extends AbstractBaseRecord> void apply( RecordStore<R> store, ProgressListener progressListener,
                 Predicate<? super R>... filters ) throws FAILURE
         {
-            for ( R record : Scanner.scan( store, true, filters ) )
+            ResourceIterable<R> iterable = Scanner.scan( store, true, filters );
+            try ( ResourceIterator<R> scan = iterable.iterator() )
             {
-                if ( shouldStop )
+                while ( scan.hasNext() )
                 {
-                    break;
-                }
+                    R record = scan.next();
+                    if ( shouldStop )
+                    {
+                        break;
+                    }
 
-                store.accept( this, record );
-                progressListener.set( record.getId() );
+                    store.accept( this, record );
+                    progressListener.set( record.getId() );
+                }
+                progressListener.done();
             }
-            progressListener.done();
         }
     }
 
     class Scanner
     {
         @SafeVarargs
-        public static <R extends AbstractBaseRecord> Iterable<R> scan( final RecordStore<R> store,
+        public static <R extends AbstractBaseRecord> ResourceIterable<R> scan( final RecordStore<R> store,
                 final Predicate<? super R>... filters )
         {
             return scan( store, true, filters );
         }
 
         @SafeVarargs
-        public static <R extends AbstractBaseRecord> Iterable<R> scan( final RecordStore<R> store,
+        public static <R extends AbstractBaseRecord> ResourceIterable<R> scan( final RecordStore<R> store,
                 final boolean forward, final Predicate<? super R>... filters )
         {
-            return () -> new PrefetchingIterator<R>()
+            return () -> new PrefetchingResourceIterator<R>()
             {
                 final PrimitiveLongIterator ids = new StoreIdIterator( store, forward );
-                final R record = store.newRecord();
+                final RecordCursor<R> cursor = store.newRecordCursor( store.newRecord() );
+                {
+                    store.placeRecordCursor( 0, cursor, RecordLoad.CHECK );
+                }
```

Comment:
I think the code will be clearer if you extracted this inner class, and moved this initialisation to it's constructor.